source/MRMesh/MRVoxelsConversions.cpp

```diff
@@ -474,74 +489,185 @@ TriangulationPlan{0, 3, 8},
+bool findSeparationPoint( SeparationPoint& sp, const VdbVolume& volume, const ConstAccessor& acc,
+                          const openvdb::Coord& coord, const Vector3i& basePos, float valueB, NeighborDir dir,
+                          const VolumeToMeshParams& params )
 {
-    auto basePos = indexer.toPos( base );
     if ( basePos[int( dir )] + 1 >= volume.dims[int( dir )] )
-        return {};
-    auto coord = openvdb::Coord{ basePos.x + minCoord.x(),basePos.y + minCoord.y(),basePos.z + minCoord.z() };
-    const auto& valueB = acc.getValue( coord );// volume.data[base];
+        return false;
     auto nextCoord = coord;
     nextCoord[int( dir )] += 1;
-    const auto& valueD = acc.getValue( nextCoord );// volume.data[nextId];
+    float valueD = acc.getValue( nextCoord );// volume.data[nextId];
+
     bool bLower = valueB < params.iso;
     bool dLower = valueD < params.iso;
-
     if ( bLower == dLower )
-        return {};
+        return false;
 
-    SeparationPoint res;
     Vector3f coordF = Vector3f( float( coord.x() ), float( coord.y() ), float( coord.z() ) );
     Vector3f nextCoordF = Vector3f( float( nextCoord.x() ), float( nextCoord.y() ), float( nextCoord.z() ) );
     auto bPos = params.origin + mult( volume.voxelSize, coordF );
     auto dPos = params.origin + mult( volume.voxelSize, nextCoordF );
-    res.position = params.positioner( bPos, dPos, valueB, valueD, params.iso );
-    res.vid = VertId{ 0 };
-    return res;
+    if constexpr ( boost::mp11::mp_contains<Args, UseDefaultVoxelPointPositioner>::value )
+        sp.position = voxelPositionerLinearInline( bPos, dPos, valueB, valueD, params.iso );
+    else
+        sp.position = params.positioner( bPos, dPos, valueB, valueD, params.iso );
+    return true;
 }
 
-SeparationPoint findSeparationPoint( const SimpleVolume& volume,
-    const VolumeIndexer& indexer, VoxelId base, NeighborDir dir, const VolumeToMeshParams& params )
+template <typename Args>
+bool findSeparationPoint( SeparationPoint& sp, const SimpleVolume& volume, const VolumeIndexer& indexer, VoxelId base,
+                          const Vector3i& basePos, NeighborDir dir, const VolumeToMeshParams& params )
 {
-    auto basePos = indexer.toPos( base );
     auto nextPos = basePos;
     nextPos[int( dir )] += 1;
     if ( nextPos[int( dir )] >= volume.dims[int( dir )] )
-        return {};
-    const auto& valueB = volume.data[base];
-    const auto& valueD = volume.data[indexer.toVoxelId( nextPos ).get()];
-    if ( std::isnan( valueB ) || std::isnan( valueD ) )
-        return {};
+        return false;
+
+    OutEdge outEdge;
+    switch ( dir )
+    {
+    case MR::MarchingCubesHelper::NeighborDir::X:
+        outEdge = OutEdge::PlusX;
+        break;
+    case MR::MarchingCubesHelper::NeighborDir::Y:
+        outEdge = OutEdge::PlusY;
+        break;
+    case MR::MarchingCubesHelper::NeighborDir::Z:
+        outEdge = OutEdge::PlusZ;
+        break;
+    default:
+        return false;
+    }
```

Comment:
Can we replace this complex switch with some simple formula like `outEdge = (OutEdge)dir`?